pvlib/ivtools.py

```diff
@@ -262,6 +262,144 @@ def fit_sde_sandia(voltage, current, v_oc=None, i_sc=None, v_mp_i_mp=None,
                                      v_oc)
 
 
+def fit_sdm_desoto(v_mp, i_mp, v_oc, i_sc, alpha_sc, beta_voc,
+                   cells_in_series, EgRef=1.121, dEgdT=-0.0002677,
+                   temp_ref=25, irrad_ref=1000):
+    """
+    Calculates the parameters for the De Soto single diode model using the
+    procedure described in [1]. This procedure has the advantage of
+    using common specifications given by manufacturers in the
+    datasheets of PV modules.
+
+    The solution is found using the scipy.optimize.root() function,
+    with the corresponding default solver method 'hybr'.
+    No restriction is put on the fit variables, i.e. series
+    or shunt resistance could go negative. Nevertheless, if it happens,
+    check carefully the inputs and their units; alpha_sc and beta_voc are
+    often given in %/K in manufacturers datasheets and should be given
+    in A/K and V/K here.
+
+    The parameters returned by this function can be used by
+    pvsystem.calcparams_desoto to calculate the values at different
+    irradiance and cell temperature.
+
+    Parameters
+    ----------
+    v_mp: float
+        Module voltage at the maximum-power point at reference conditions [V].
+    i_mp: float
+        Module current at the maximum-power point at reference conditions [A].
+    v_oc: float
+        Open-circuit voltage at reference conditions [V].
+    i_sc: float
+        Short-circuit current at reference conditions [A].
+    alpha_sc: float
+        The short-circuit current (i_sc) temperature coefficient of the
+        module [A/K].
+    beta_voc: float
+        The open-circuit voltage (v_oc) temperature coefficient of the
+        module [V/K].
+    cells_in_series: float
+        Number of cell in the module.
+    EgRef: float, default 1.121 eV - value for silicon
+        Energy of bandgap of semi-conductor used [eV]
+    dEgdT: float, default -0.0002677 - value for silicon
+        Variation of bandgap according to temperature [eV/K]
+    temp_ref: float, default 25
+        Reference temperature condition [C]
+    irrad_ref: float, default 1000
+        Reference irradiance condition [W/m2]
+
+    Returns
+    -------
+    Tuple of the following elements:
+
+        * Dictionary with the following elements:
```

Comment:
I agree that it’s good to be consistent. However, I very much prefer interfaces where order does _not_ matter, because this ordering is a form of complexity, which is harder to maintain/extend without breakages. Also, the dictionary output is perhaps _best_ paired with keyword-only inputs to subsequent functions/methods, but it still works with pvlib’s current mixture of named positional and keyword arguments. However, it would not be compatible with Python 3.8’s new positional-only arguments.

Reply:
I don't think there's a right answer. It would be most helpful if people could provide some good references for pros/cons of modern scientific python design patterns. I am not aware of any references that tackle non-trivial cases.

Reply:
24 minutes into this keynote talk on simplicity, Rich Hickey (creator of Clojure) discusses the complexity of ordering things in lists:

https://m.youtube.com/watch?v=rI8tNMsozo0&t=3s

This is not gospel of course, and indeed ordering can provide performance gains, etc. (which is why dictionaries are now ordered in CPython, which I personally treat as an implementation detail to not to be relied upon).

Reply:
Let's address consistency among functions returns in a follow-on PR. I can't say I had a pattern in mind when I started `ivtools.py`